### PR TITLE
Add file uploads support

### PR DESCRIPTION
**Pull request by Ener-Getick**
This PR add file uploads support. (#793)
Currently, there are no requirements supported.

 ```php
<?php

use FOS\RestBundle\Controller\Annotations\QueryParam;

/**
 * @FileParam(
 *   name="foo_file",
 *   key=null,
 *   incompatibles={},
 *   default=null,
 *   description="",
 *   strict=false,
 *   nullable=false,
 *   requirements={},
 *   image=false
 * )
 */
```

I moved the scalar constraints (notBlank, array) in another abstract class ``ScalarParam``, I don't know if this is a really good name ...

The ``requirements`` parameter of ``@FileParam`` must be a constraint or an array that can contain all the fields of http://symfony.com/doc/current/reference/constraints/File.html or http://symfony.com/doc/current/reference/constraints/Image.html. It depends of the ``image`` parameter.

### CODE REVIEW

**Comment by lsmith77** [comment]
please rebase on master

**Comment by lsmith77** [comment]
can you also add something to the docs?

**Comment by Ener-Getick** [comment]
Something like that ?

File: Request/ParamFetcher.php [hunk 1]
```diff
@@ -155,20 +162,22 @@ public function get($name, $strict = null)
...
                         sprintf('%s parameter "%s" is %s', $paramType, $name, $problem)
                     );
                 }
-
-                return $this->cleanParamWithRequirements($config, $param, $strict);
+            } else {
+                return $default;
             }
-
-            return $default;
```
Comment (lsmith77) [inline Request/ParamFetcher.php]: and what happens now in this case .. we return null?

File: Tests/EventListener/VersionListenerTest.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+use FOS\RestBundle\View\View;
+use FOS\RestBundle\View\ViewHandler;
+
+/**
+ * Version listener test.
+ *
+ * @author Ener-Getick <egetick@gmail.com>
+ */
+class VersionListenerTest extends \PHPUnit_Framework_TestCase
```
Comment (lsmith77) [inline Tests/EventListener/VersionListenerTest.php]: I guess this is an unrelated addition to the test suite .. right?

File: Request/ParamFetcher.php [hunk 3]
```diff
@@ -11,15 +11,21 @@
```
Comment (lsmith77) [inline Request/ParamFetcher.php]: I am wondering if we do not need to refactor this a bit to bring the methods back down to a manageable level of complexity .. quite some deep if constructs we have here ..

File: Request/ParamFetcher.php (hunk 1 above)
Comment (Ener-Getick) [inline Request/ParamFetcher.php]: Look the code, it's the same thing, I just removed an useless call to simplify.

File: Tests/EventListener/VersionListenerTest.php (hunk 2 above)
Comment (Ener-Getick) [inline Tests/EventListener/VersionListenerTest.php]: Yes in part, I'm trying to understand how work the version listener to make this PR compatible.
Do you want me to split this change in another commit / PR ?

File: Request/ParamFetcher.php (hunk 3 above)
Comment (Ener-Getick) [inline Request/ParamFetcher.php]: It's probably a good idea, the ``ParamFetcher`` begins to be really complex...
Do you want me to include #1075 in this PR ?

**Comment by Ener-Getick** [comment]
I refactored the ``ParamFetcher``. Most of the logic is moved into the ``ParamInterface`` annotations.

It's now easier to fix some issues like #799.

**Comment by lsmith77** [comment]
needs a rebase

**Comment by Ener-Getick** [comment]
done.

**Comment by lsmith77** [comment]
and another one :-/

**Comment by Ener-Getick** [comment]
done :-)

File: Controller/Annotations/AbstractParam.php [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    public $nullable = false;
+    /** @var array */
+    public $incompatibles = array();
+
+    /** {@inheritdoc} */
+    public function getName()
+    {
+        return $this->name;
+    }
+    /** {@inheritdoc} */
```
Comment (lsmith77) [inline Controller/Annotations/AbstractParam.php]: please add an empty line between methods

File: Controller/Annotations/AbstractScalarParam.php [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+{
+    /** @var scalar|array|Constraint|null */
+    public $requirements = null;
+    /** @var bool */
+    public $array = false;
+    /** @var bool */
+    public $allowBlank = true;
+
+    /** {@inheritdoc} */
+    public function getConstraints() {
```
Comment (lsmith77) [inline Controller/Annotations/AbstractScalarParam.php]: ``{`` needs to be on the next line

File: Controller/Annotations/AbstractScalarParam.php [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        $constraints = parent::getConstraints();
+
+        $constraints[] = new Constraints\Type(array('type' => 'scalar'));
+        $requirements = $this->resolve($this->requirements);
+
+        if($this->requirements instanceof Constraint) {
```
Comment (lsmith77) [inline Controller/Annotations/AbstractScalarParam.php]: space missing

File: Controller/Annotations/AbstractScalarParam.php [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+            $constraints[] = $requirements;
+        } elseif(is_scalar($requirements)) {
```
Comment (lsmith77) [inline Controller/Annotations/AbstractScalarParam.php]: space missing

File: Controller/Annotations/AbstractScalarParam.php [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+            $constraints[] = new Constraints\Regex(array(
+                'pattern' => '#^(?:'.$requirements.')$#xsu',
+                'message' => sprintf(
+                    "Parameter '%s' value, does not match requirements '%s'",
+                    $this->getName(),
+                    $requirements
+                ),
+            ));
+        } elseif(is_array($requirements) && isset($requirements["rule"]) && $requirements["error_message"]) {
```
Comment (lsmith77) [inline Controller/Annotations/AbstractScalarParam.php]: space missing and please use single quotes where possible

File: Controller/Annotations/AbstractScalarParam.php [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+                'message' => $requirements["error_message"]
+            ));
+        }
+
+        if(false === $this->allowBlank) {
+            $constraints[] = new Constraints\NotBlank();
+        }
+
+        // If an array is expected apply the constraints to each element.
+        if($this->array) {
```
Comment (lsmith77) [inline Controller/Annotations/AbstractScalarParam.php]: space missing

File: Controller/Annotations/AbstractScalarParam.php [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+                ),
+            ));
+        } elseif(is_array($requirements) && isset($requirements["rule"]) && $requirements["error_message"]) {
+            $constraints[] = new Constraints\Regex(array(
+                'pattern' => '#^(?:'.$requirements["rule"].')$#xsu',
+                'message' => $requirements["error_message"]
+            ));
+        }
+
+        if(false === $this->allowBlank) {
```
Comment (lsmith77) [inline Controller/Annotations/AbstractScalarParam.php]: space missing

File: Controller/Annotations/FileParam.php [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+        $options = $this->requirements ?: array();
+        if ($this->image) {
+            $constraints[] = new Constraints\Image($options);
+        } else {
+            $constraints[] = new Constraints\File($options);
+        }
+
+        return $constraints;
+    }
+    /**
```
Comment (lsmith77) [inline Controller/Annotations/FileParam.php]: please add one empty line here

**Comment by lsmith77** [comment]
actually .. I just activated styleci ..

https://github.com/FriendsOfSymfony/FOSRestBundle/pull/1115

**Comment by Ener-Getick** [comment]
I'll wait that you merge #1115 to see if there is some standards mistakes related to this PR.

**Comment by lsmith77** [comment]
indeed .. needs another rebase :-/

form now on it should then also show you any CS violations automatically

**Comment by Ener-Getick** [comment]
Yes, I fixed them.

**Comment by lsmith77** [comment]
awesome!

**Comment by ReservedDeveloper** [comment]
In attempting to test this out (specifically using the example with `@QueryParam(name="page", requirements="\d+", default="1", description="Page of the overview.")`, I ran in to an issue with the annotations parser. Specifically [Doctrine/Common/Annotations/DocParser.php:786](https://github.com/doctrine/annotations/blob/master/lib/Doctrine/Common/Annotations/DocParser.php#L786):

```php
} elseif (gettype($values[$property]) !== $type['type'] && !$values[$property] instanceof $type['type']) {
```
Note the `gettype` call when `$values[$property]` is a string will return `string` type, when only `scalar|array|Constraint|null` are supported.

Since string !== (scalar|array|Constraint|null)... :boom:

The specific error I'm receiving is:
```sh
  [Doctrine\Common\Annotations\AnnotationException]                                                                                                                                                     
  [Type Error] Attribute "requirements" of @FOSRest\QueryParam declared on method AppBundle\Controller\XyzController::indexAction() expects a(n) scalar|array|Constraint|null, but got string.
```

I didn't see any changes to the dependencies that would suggest a different version of the Doctrine library. Am I missing something here?

**Comment by Ener-Getick** [comment]
Hum so it seems that the doctrine annotations reader doesn't support the scalar type... If you have the time, can you change this doc and add a test for this case please? 

Instead I'll fix this later, probably tomorrow.

**Comment by xabbuh** [comment]
@Ener-Getick @lsmith77 For some reason it seems that this change is not in the `1.8` branch. Do you now if that was intended?

**Comment by Ener-Getick** [comment]
@xabbuh Yeah it needs a lot of changes to be backported this is why I didn't do it.

**Comment by xabbuh** [comment]
@Ener-Getick Alright, thanks for the confirmation.

**Comment by Ener-Getick** [comment]
But at least we can backport the ``map`` option in parameters. It should be pretty simple.
